README.md
# IoT FabLab Order Management
[![Build Status](https://travis-ci.com/FabLabLuebeckEV/Print4Life.svg?branch=develop)](https://travis-ci.com/FabLabLuebeckEV/Print4Life)
## Requirements

* Node > 12.13.1
* MongoDB (or at least a server with it)
	* Locally:
	```bash
	docker run --rm -d \
	--name order-mongo \
	-v /path/to/git/order-management/dev-mongo:/data/db \
	-p 27017:27017 \
	mongo
	```
	* Server:
	`ssh -L 27017:localhost:27017 remote-user@212.83.56.107`
* Webserver (Nginx, Apache) for production (Let's Encrypt for Certificate)
* PM2

### Setup

* Install Node
* Install Mongo
* Install PM2 via npm
* Install a Webserver with SSL Certificate
* Check Configuration of this project (order-management/api/config/config.ts for backend and order-management/frontend/app/config/routes.ts for frontend)

#### IBM  Watson Setup

* Create an IBM Watson IoT Platform organization
* Create a role and remember it's name, the role has to be able to create api keys
* Create a device type named "Sensor"
* Create an application to get an API Key and it's password
* Update: ecosystem-*, config.ts, environment* with OrgID (top right corner on every page), API Key and API password

If iot-tests fail, try deleting existing testcase-users as there might be an incorrect entry

## Use Node-Debugger with VSCode

If you use this boilerplate with VSCode you can run the debugger for node with ```npm run api-dev``` and use the feature in the bottom line 'auto attach' when running 'Attach to NPM'. This auto attaches the debugger to the running npm process.

## Run the app

To run the app in dev mode you need to run ```npm run api-dev``` and ```npm run ng-dev```. The application will then run on localhost:4200 for the angular app and localhost:3000 for the node backend. These commands will rebuild the nodets and ng application on changed files. **Beaware that the backend server only starts when a debugger is attached to it**.

### Admin User

You can login to the initial admin user with *username:* **admin** and *password:* **E9bBs8UpeASq**


CHANGELOG.md
# Changelog

## Unreleased / Upcoming next

* multilanguage support for English, German and Danish (/LF0380/)

### No Version number yet

### Version 0.8
* upgrade from Angular 6 to 8
* implemented new Design
* implemented print4life-idea (makers and hospitals)
* switch login to email and password

### Version 0.7
* users can activate themselfs by a link via email
* landing page redsigned and added support for mobile devices
* public orders heavily redisgned
* initial location will be retrieved from the browsers
* maschine detail show correct schedules
* no more orders without empty user addresses
* user address is only visible for the user itself and the admins
* fixed a lot of typos and measurment units
* fixed input types for various input fields

### Version 0.6.2

* correctly parse env variable NG_PORT as number and therefore correctly check for it being port 80 to set baseURLFrontend with or without port addition
* implement new env variable for EMAIL_ADDRESS as it may differ from EMAIL_USER
* add ecosystem for tzl deploy
* add log message for http or https server start to see if the server accepted SSL keys

### Version 0.6.1

* fix table headers of order list (now shows on empty table)
* use baseUrlFrontend as whitelist for cors
* set email secure to false for production and add another api password for gmail
* show table headers even for empty tables
* update table of iot device list view if a device is deleted
* disable machine info button on create shared order
* set validation error messages for shipping address fields of create order form
* correct behaviour of scheduling validation of create order form (now allows only complete start- and end dates)
* schedules are deleted if they are emptied on create order form (former: didn't delete model relation with order <--> schedule)
* make loggedInUser a component variable on order list

### Version 0.6

* added ModalService
* added Buttons as Components
* added IoT Device Support (/LF0010/, /LF0020/, /LF0030/, /LF0040/)
* added Octoprint bridge
* fixed DatePicker ChangeEvent on entering input without DatePicker
* unfinished orders filters fixed on schedules and fablabs
* fixed spinner on order list
* tooltips on buttons

### Version 0.5.2

* email fix (smtp config updated)
* added fablabs in user list and detail view
* filter for fablabs added in user list
* fix error messages to use old and new styled error messages

### Version 0.5.1

* fix cors to enable non port urls
* add timePickError as reason why to disable the submit button in create order view

### Version 0.5

* implemented upload of files (/LF0300/, /LF0330/)
* implemented unfinsihed orders list (/LF0270/)
* implemented success history for orders (/LF0280/)
* implemented machine statistics (currently successful orders; /LF0280/)
* implemented shared orders to allow guests to create, edit and view orders (/LF0360/)
* added shipping address to orders (selectable via user address, custom address, fablab (of user) address)
* refactored backend controller to use own services
* routes decoupled from controller functions
* added search for username, firstname, lastname and email of users in user list
* added search for projectname of orders in order lists
* added search for devicename and manufacturer of machine in machine lists
* added scheduling for orders and machines
* deleting machines results to state deactivate instead of a hard deletion
* added deployment to production server
* create oder form now uses local storage to contain some information on changing views
* adding checkbox for copyright question on uploaded files
* adding unknown machine type on create order for users that don't know the correct machine type
* order form now only shows machines of a specific fablab (choosing a fablab is a mandatory field now)
* order form now only shows editors of a specific fablab
* download of files is only allowed for owner, editors and admins
* files are now deletable and deprecated files are deleted if order status switches to 'completed', 'representive', 'archived' or 'deleted'
* unfinished order list can now be filtered by machine types

### Version 0.4.1

* fix cors with https
* fix specs and use env configs
* use ecosystem.config.js file for pm2
* setup https backend server for prod
* fix deploy scripts to include email-templates in dist build
* fix duplicate orders in order list
* fix user based switch language

### Version 0.4

* user and role management implemented
* users can register to the platform (/LF0210/)
* users can now login (/LF0220/)
* admins receive emails regarding new registered users and activation requests of inactive user profiles
* users can reset passwords if they can't remember their password via entering the e-mail address. If the e-mail address is correct, a new password will be sent to that address (/LF0240/)
* users can change passwords on their profile page (/LF0330/)
* users can select a preferred language which will be used on the frontend and also for received emails (no dk support yet)
* users can only edit own orders (except for editors/admins)
* orders are assignable to editors (/LF0350/)
* comments, owner and editors are now based on user objects instead of strings
* fablab is now an entity with an address assignable to users as an optional field
* renaming printers to 3d-printers
* guests (not logged in users) now can only see the order list

## Version 0.3.2

* translation fix in create order form

## Version 0.3.1

* minor cors fix

## Version 0.3

* implemented translation to English and German

## Version 0.2.1

* add dns to cors settings
* add changelog

## Version 0.2

* basic CRUDs for machines
* basic CRUDs for orders
* frontend to list machines (/LF0250/)
* frontend to list orders (/LF0260/)
* frontend to create, edit, read and delete machines (not included upload of files and CRUD for laser types and material; regarding spec /LF0370/)

api/config/config.ts
const env = process.env.NODE_ENV || 'dev';
const port = process.env.PORT || 3000;
const ngPort = process.env.NG_PORT ? Number.parseInt(process.env.NG_PORT, 10) : 4200;
const ibmWatson = {
  key: process.env.WATSON_API_KEY || 'a-o7bxcz-hgc8qykrqj',
  token: process.env.WATSON_API_PASSWORD || 'gi)0-7UQ-QEm_f+k7J',
  orgId: process.env.WATSON_ORG_ID || 'o7bxcz',
  userRoles: ['PD_STANDARD_APP']
};
const jwtSecret = env === 'dev' || env === 'test' || env === 'testLocal'
  ? 'phahng9tie6uthashe4Deng8Iek0eefahv9aawu1ah'
  : process.env.JWT_SECRET;

const baseUrl = '/api/v1/';
const baseUrlBackend = `${process.env.BASE_URL_BACKEND || 'http://localhost'}:${port}${baseUrl}`;
const baseUrlFrontend = ngPort === 80
  ? `${process.env.BASE_URL_FRONTEND || 'http://localhost'}`
  : `${process.env.BASE_URL_FRONTEND || 'http://localhost'}:${ngPort}`;
/**
 * set to 2 hours
 * (values without unit are by default ms,
 * see https://github.com/auth0/node-jsonwebtoken#jwtsignpayload-secretorprivatekey-options-callback)
 */
const jwtExpiryTime = '2h';
const attachmentBucket = 'orderAttachments';

const tmpDir = process.env.TMP_DIR || './tmp';
const email = {
  host: process.env.EMAIL_HOST || 'smtp.gmail.com',
  port: process.env.EMAIL_PORT || 587,
  secure: process.env.EMAIL_SECURE === 'true',
  from: process.env.EMAIL_ADDRESS,
  auth: {
    user: process.env.EMAIL_USER || 'iot.fablab@gmail.com',
    pass: process.env.EMAIL_PASS || 'scdkhsxiumwwgiob'
  }
};

const publicRoutes = [
  {
    url: `${baseUrl}orders`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}orders/:id`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}orders/count`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}orders/status`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}orders/status/outstanding`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}orders/search`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}orders/shared/:id/upload`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}orders/shared/:id/comment`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}orders/shared/:id/download/:fileId/`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}users/login`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}users/:id/activationRequest`,
    methods: ['PUT']
  },
  {
    url: `${baseUrl}users/:id/getNames`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}users/`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}users/resetPassword/`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}users/roles`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}users/languages`,
    methods: ['GET']
  },
  {
    url: `${baseUrl}users/:id/activate`,
    methods: ['PUT']
  },
  {
    url: `${baseUrl}machines`,
    canChilds: true,
    methods: ['GET']
  },
  {
    url: `${baseUrl}machines/3d-printers/count`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/lasercutters/count`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/millingMachines/count`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/otherMachines/search`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/3d-printers/search`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/lasercutters/search`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/millingMachines/search`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}machines/otherMachines/count`,
    methods: ['POST']
  },
  {
    url: `${baseUrl}fablabs`,
    canChilds: true,
    methods: ['GET']
  }
];
const dev = {
  jwtSecret,
  jwtExpiryTime,
  attachmentBucket,
  ibmWatson,
  tmpDir,
  ssl: {
    privateKeyPath: process.env.SSL_PRIV_KEY || '',
    certificatePath: process.env.SSL_CERT || ''
  },
  loggerRotateOptions: {
    datePattern: 'DD-MM-YYYY',
    dirname: 'logs',
    maxSize: '5m',
    maxFiles: '10'
  },
  email,
  connections: {
    mongo: {
      host: process.env.MONGO_HOST || 'mongodb://127.0.0.1:27017/',
      database: process.env.MONGO_DBNAME || 'iot-fablab-dev'
    }
  },
  rawBaseUrl: baseUrl,
  publicRoutes,
  baseUrlBackend,
  baseUrlFrontend,
  cors: {
    whitelist: [baseUrlFrontend]
  }
  // cors: undefined // if testing backend routes in dev mode without using the frontend
};

const staging = {
  jwtSecret,
  jwtExpiryTime,
  attachmentBucket,
  ibmWatson,
  tmpDir,
  ssl: {
    privateKeyPath: process.env.SSL_PRIV_KEY,
    certificatePath: process.env.SSL_CERT
  },
  loggerRotateOptions: {
    datePattern: 'DD-MM-YYYY',
    dirname: 'logs',
    maxSize: '10m',
    maxFiles: '30'
  },
  email,
  connections: {
    mongo: {
      host: process.env.MONGO_HOST,
      database: process.env.MONGO_DBNAME
    }
  },
  rawBaseUrl: baseUrl,
  publicRoutes,
  baseUrlBackend,
  baseUrlFrontend,
  cors: {
    whitelist: [
      baseUrlFrontend
    ]
  }
  // cors: undefined
};

const prod = {
  jwtSecret,
  jwtExpiryTime,
  attachmentBucket,
  ibmWatson,
  tmpDir,
  ssl: {
    privateKeyPath: process.env.SSL_PRIV_KEY,
    certificatePath: process.env.SSL_CERT
  },
  loggerRotateOptions: {
    datePattern: 'DD-MM-YYYY',
    dirname: 'logs',
    maxSize: '10m',
    maxFiles: '30'
  },
  email,
  connections: {
    mongo: {
      host: process.env.MONGO_HOST,
      database: process.env.MONGO_DBNAME
    }
  },
  rawBaseUrl: baseUrl,
  publicRoutes,
  baseUrlBackend,
  baseUrlFrontend,
  cors: {
    whitelist: [
      baseUrlFrontend
    ]
  }
  // cors: undefined
};

const test = {
  jwtSecret,
  jwtExpiryTime,
  attachmentBucket,
  ibmWatson,
  tmpDir,
  ssl: {
    privateKeyPath: process.env.PRIV_KEY || '',
    certificatePath: process.env.CERT || ''
  },
  loggerRotateOptions: {
    datePattern: 'DD-MM-YYYY',
    dirname: 'logs',
    maxSize: '5m',
    maxFiles: '5'
  },
  email,
  connections: {
    mongo: {
      host: process.env.MONGO_HOST || 'mongodb://mongo:27017/',
      database: process.env.MONGO_DBNAME || 'iot-fablab'
    }
  },
  rawBaseUrl: baseUrl,
  publicRoutes,
  baseUrlBackend: `${process.env.BASE_URL_BACKEND || 'http://localhost'}:${port}${baseUrl}`,
  baseUrlFrontend: `${process.env.BASE_URL_FRONTEND || 'http://localhost'}:${ngPort}`,
  cors: undefined
};

const testLocal = {
  jwtSecret,
  jwtExpiryTime,
  attachmentBucket,
  ibmWatson,
  tmpDir,
  ssl: {
    privateKeyPath: process.env.PRIV_KEY || '',
    certificatePath: process.env.CERT || ''
  },
  loggerRotateOptions: {
    datePattern: 'DD-MM-YYYY',
    dirname: 'logs',
    maxSize: '5m',
    maxFiles: '5'
  },
  email,
  connections: {
    mongo: {
      host: process.env.MONGO_HOST || 'mongodb://127.0.0.1:27017/',
      database: process.env.MONGO_DBNAME || 'iot-fablab-test'
    }
  },
  rawBaseUrl: baseUrl,
  publicRoutes,
  baseUrlBackend: `${process.env.BASE_URL_BACKEND || 'http://localhost'}:${port}${baseUrl}`,
  baseUrlFrontend: `${process.env.BASE_URL_FRONTEND || 'http://localhost'}:${ngPort}`,
  cors: undefined
};

export const configArr = {
  dev, prod, test, testLocal, staging
};

const config = configArr[env];

export default config;
